index.js

```diff
@@ -0,0 +1,316 @@
+                        require ('sanctuary-show'),
+                        require ('sanctuary-type-classes'));
+  } else if (typeof define === 'function' && define.amd != null) {
+    define (['sanctuary-show', 'sanctuary-type-classes'], function(show, Z) {
+      return f (util, show, Z);
+    });
+  } else {
+    self.sanctuaryConstant = f (util,
+                                self.sanctuaryShow,
+                                self.sanctuaryTypeClasses);
+  }
+
+} (function(util, show, Z) {
+
+  'use strict';
+
+  /* istanbul ignore if */
+  if (typeof __doctest !== 'undefined') {
+    var $ = __doctest.require ('sanctuary-def');
+    var type = __doctest.require ('sanctuary-type-identifiers');
+    var S = (function() {
+      var S = __doctest.require ('sanctuary');
+      var ConstantType = $.BinaryType
+        ('Constant')
+        ('')
+        ([])
+        (function(x) { return type (x) === constantTypeIdent; })
+        (function(c) { return [c.value]; })
+        (function(c) { return []; });
+      var env = Z.concat (S.env, [ConstantType ($.Unknown) ($.Unknown)]);
+      return S.create ({checkTypes: false, env: env});
+    } ());
+  }
+
+  var constantTypeIdent = 'sanctuary-constant/Constant@1';
+
+  //. `Constant a b` satisfies the following [Fantasy Land][] specifications:
+  //.
+  //. ```javascript
+  //. > const Useless = require ('sanctuary-useless')
+  //.
+  //. > S.map (k =>
+  //. .   k + ' '.repeat (16 - k.length) +
+  //. .   (Z[k].test (Constant (Useless.constructor) (Useless)) ? '\u2705   ' :
+  //. .    Z[k].test (Constant (Array) (['foo', 'bar', 'baz'])) ? '\u2705 * ' :
+  //. .    /* otherwise */                                        '\u274C   ')
+  //. . ) (S.filter (S.test (/^(?=[A-Z])(?!TypeClass$)/)) (Object.keys (Z)))
+  //. [ 'Setoid          ✅ * ',  // if ‘a’ satisfies Setoid
+  //. . 'Ord             ✅ * ',  // if ‘a’ satisfies Ord
+  //. . 'Semigroupoid    ❌   ',
+  //. . 'Category        ❌   ',
+  //. . 'Semigroup       ✅ * ',  // if ‘a’ satisfies Semigroup
+  //. . 'Monoid          ❌   ',
+  //. . 'Group           ❌   ',
+  //. . 'Filterable      ❌   ',
+  //. . 'Functor         ✅   ',
+  //. . 'Bifunctor       ✅   ',
+  //. . 'Profunctor      ❌   ',
+  //. . 'Apply           ✅ * ',  // if ‘a’ satisfies Semigroup
+  //. . 'Applicative     ✅ * ',  // if ‘a’ satisfies Monoid
+  //. . 'Chain           ❌   ',
+  //. . 'ChainRec        ❌   ',
+  //. . 'Monad           ❌   ',
+  //. . 'Alt             ❌   ',
+  //. . 'Plus            ❌   ',
+  //. . 'Alternative     ❌   ',
+  //. . 'Foldable        ✅   ',
+  //. . 'Traversable     ✅   ',
+  //. . 'Extend          ❌   ',
+  //. . 'Comonad         ❌   ',
+  //. . 'Contravariant   ❌   ' ]
+  //. ```
+
+  //# Constant :: TypeRep a -> a -> Constant a b
+  //.
+  //. Constant's sole data constructor. The [type representative][] makes
+  //. `Constant (M)` an [Applicative][]-compatible type representative if
+  //. `M` represents some monoidal type.
```

Comment:
Is this helpful, @Avaq?

Reply:
I read this a few times yesterday, but without being able to think up a practical use-case, I couldn't understand what it's for.

Reply:
@Avaq Imagine you're trying to derive `foldMap :: Monoid m -> (a -> m) -> f a -> m` from `traverse :: Applicative g -> (a -> g b) -> f a -> g (f b)`. So you start writing:

```
const foldMap = ({ empty, append }) => am => fa => {
  // ?
}
```

Now obviously the only tool you have available is `traverse`, from which you're trying to derive `foldMap`, so you write:

```
const foldMap = ({ empty, append }) => am => fa => traverse(/* ?1 */)(/* ?2 */)(/* ?3 */)
```

Now you're stuck with the task of producing an `Applicative g` to feed into `?1`, an `a -> g b` to feed into `?2` and a `f a` to feed into `?3`. It just so happens we already have an `fa :: f a` we received from the arguments, so that takes care of `?3` at least:

```
const foldMap = ({ empty, append }) => am => fa => traverse(/* ?1 */)(/* ?2 */)(fa)
```

Let's think about the types again:

```
traverse :: Applicative g -> (a -> g b) -> f a -> g (f b)
foldMap  :: Monoid m      -> (a -> m)   -> f a -> m
```

we can see a few things:

- it must be the case that `g (f b) = m`
- `g` (whatever it is) needs to have an instance `Applicative g`
- this derivation must work regardless of what traversable `f` is: i.e. `f b` could be any type whatsoever
- `m` again is beyond our control: so whatever `(a -> m)` the user provides to `foldMap`, it is precisely for that `m` that we must have `g (f b) = m`

So, what's an applicative functor `g` such that `g x = m` for an arbitrary type `x` and an arbitrary monoid `m`? :wink: